tests/describe_task_queue_test.go

```diff
@@ -0,0 +1,230 @@
+	s.UpdateUtils = updateutils.New(s.T())
+}
+
+func (s *DescribeTaskQueueSuite) TestAddNoTasks_ValidateBacklogInfo() {
+	s.PublishConsumeWorkflowTasksValidateBacklogInfo(4, 0)
+}
+
+func (s *DescribeTaskQueueSuite) TestAddSingleTask_ValidateBacklogInfo() {
+	s.PublishConsumeWorkflowTasksValidateBacklogInfo(1, 1)
+}
+
+func (s *DescribeTaskQueueSuite) TestAddMultipleTasksMultiplePartitions_ValidateBacklogInfo() {
+	s.PublishConsumeWorkflowTasksValidateBacklogInfo(4, 4)
+}
+
+func (s *DescribeTaskQueueSuite) PublishConsumeWorkflowTasksValidateBacklogInfo(partitions int, workflows int) {
+	// overriding the ReadPartitions and WritePartitions
+	dc := s.testCluster.host.dcClient
+	dc.OverrideValue(s.T(), dynamicconfig.MatchingNumTaskqueueReadPartitions, partitions)
+	dc.OverrideValue(s.T(), dynamicconfig.MatchingNumTaskqueueWritePartitions, partitions)
+
+	tl := "backlog-counter-task-queue"
+	tq := &taskqueuepb.TaskQueue{Name: tl, Kind: enumspb.TASK_QUEUE_KIND_NORMAL}
+	identity := "worker-multiple-tasks"
+
+	for i := 0; i < workflows; i++ {
+
+		id := uuid.New()
+		wt := "functional-workflow-multiple-tasks"
+		workflowType := &commonpb.WorkflowType{Name: wt}
+
+		request := &workflowservice.StartWorkflowExecutionRequest{
+			RequestId:           uuid.New(),
+			Namespace:           s.namespace,
+			WorkflowId:          id,
+			WorkflowType:        workflowType,
+			TaskQueue:           tq,
+			Input:               nil,
+			WorkflowRunTimeout:  durationpb.New(20 * time.Second),
+			WorkflowTaskTimeout: durationpb.New(3 * time.Second),
+			Identity:            identity,
+		}
+
+		resp0, err0 := s.engine.StartWorkflowExecution(NewContext(), request)
+		s.NoError(err0)
+
+		we := &commonpb.WorkflowExecution{
+			WorkflowId: id,
+			RunId:      resp0.RunId,
+		}
+
+		s.EqualHistoryEvents(`
+  1 WorkflowExecutionStarted
+  2 WorkflowTaskScheduled`, s.getHistory(s.namespace, we))
+	}
+
+	response := s.ValidateDescribeTaskQueue(tl)
+	versionInfo := response.GetVersionsInfo()[""]
+
+	// verifying backlog counter increased as tasks were added
+	addedTasks := int64(workflows)
+	for qT, t := range versionInfo.GetTypesInfo() {
+		queueType := enumspb.TaskQueueType(qT)
+		if queueType == enumspb.TASK_QUEUE_TYPE_WORKFLOW {
+			s.Assert().Equal(t.BacklogInfo.ApproximateBacklogCount, addedTasks)
+		} else if queueType == enumspb.TASK_QUEUE_TYPE_ACTIVITY {
+			s.Assert().Equal(t.BacklogInfo.ApproximateBacklogCount, int64(0))
+		}
+	}
+
+	// starting and completing the workflow tasks by fetching their workflowExecutions
+	for i := 0; i < workflows; i++ {
+		resp1, err1 := s.engine.PollWorkflowTaskQueue(NewContext(), &workflowservice.PollWorkflowTaskQueueRequest{
+			Namespace: s.namespace,
+			TaskQueue: tq,
+			Identity:  identity,
+		})
+		s.NoError(err1)
+		s.Equal(int32(1), resp1.GetAttempt())
+
+		// responding with completion
+		_, err2 := s.engine.RespondWorkflowTaskCompleted(NewContext(), &workflowservice.RespondWorkflowTaskCompletedRequest{
```

Comment:
I think we should not make this call. From Matching point of view once the poller polls the task it's completed. We do not to report task completion as it may result in new tasks being generated by History server and messing up our counters.

Reply:
The reason why I added this call is because if we don't complete with `RespondWorkflowTaskCompleted`, the backlog Counter does not get decremented.

Moreover, if setting `ReturnNewWorkflowTask` to `true` and `ForceCreateNewWorkflowTask` to false should result in the poller being able to consecutively do the original tasks itself, which eventually results in the backlog decrementing.

Thoughts?

Reply:
RespondWorkflowTaskCompleted does not have anything to do with backlog tasks. it's for workflows progress, hence potentially creating more tasks.
Backlog counter decrements when we dispatch task (as part of processing the poll request)
Not sure what you mean by "consecutively do the original tasks itself" but if RespondWorkflowTaskCompleted returns a task it's going to be a new task, not the same task.

Reply:
I seemed to have confused wf execution with task dispatching. My sincere apologies.
Changes have been made and I agree with your suggestion!